modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_remote_downsampling_strategy.py

```diff
@@ -62,6 +62,16 @@ def __init__(
         # CoresetSupportingModule for model implementations.
         self.requires_coreset_supporting_module = False
 
+        # Some methods might not need information from forward pass (e.g. completely random)
+        # Most do (definition), hence we default to True
+        # We might want to refactor those downsamplers to presamplers and support some
+        # adaptivity at the selector, but for now we allow random downsamplers mostly
+        # to support RS2.
+        self.forward_required = True
+
+        # Some methods might only support StB, not BtS.
+        self.supports_bts = True
```

Comment:
I think we should have this. The default is bts, and if for some reason some downsampling policy does not support it, it can override the field.

The problem is that the pydantic validation happens on the user side of things. But if we at some future point decide to use the downsampling code in other contexts (similar to how we use the training code now in RHO-LOSS), then suddenly user validations do not hold anymore. Since BTS RS2 is undefined behavior, I'd rather fail here then have silent errors that one does not notice

Reply:
I understand your concern, but I like [defensive programming](https://en.wikipedia.org/wiki/Defensive_programming). I think it's a tradeoff between complexity and validating internal data. It's like using `>=` as a loop exit condition when actually you think only `==` is going to happen - are you _really_ sure that you variable will not for whatever reason increase by 2 within a single iteration?

I'd add the bts flag when the time comes that such a downsampler exists :D I am not saying we should add lots of assertions that are unnecessary, but where we do not have complete trust, I'd add them. The trainer is receiving input that goes via gRPC and whatnot - we have no guarantee (from the trainer perspective) that the input we got is validated. We asssume only the supervisor is going to send the trainer data here but who knows what will happen, anybody could send a serialized gRPC request.

TL;DR I would like to keep it